flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/component/AbstractUserClassPathJobGraphRetrieverTest.java

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.entrypoint.component;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests for the {@link AbstractUserClassPathJobGraphRetriever}.
+ */
+public class AbstractUserClassPathJobGraphRetrieverTest extends TestLogger {
+
+	@Rule
+	public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+	private static class TestJobGraphRetriever extends AbstractUserClassPathJobGraphRetriever {
+		public TestJobGraphRetriever(String jobDir) {
+			super(jobDir);
+		}
+
+		@Override
+		public JobGraph retrieveJobGraph(Configuration configuration) {
+			return null;
+		}
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testGetUserClassPathFromAFile() throws IOException {
+		final String fileName = "a.jar";
+		File file = temporaryFolder.newFile(fileName);
+
+		TestJobGraphRetriever testJobGraphRetriever = new TestJobGraphRetriever(file.getAbsolutePath());
+
+		testJobGraphRetriever.getUserClassPaths();
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testGetUserClassPathFormADoesNotExistsFile() throws IOException {
+		final String fileName = "_does_not_exists_file";
+		final String doesNotExistsFilePath = temporaryFolder.getRoot() + "/" + fileName;
+
+		TestJobGraphRetriever testJobGraphRetriever = new TestJobGraphRetriever(doesNotExistsFilePath);
+
+		assertEquals(Collections.emptyList(), testJobGraphRetriever.getUserClassPaths());
+	}
+
+	@Test
+	public void testGetUserClassPath() throws IOException {
+		final Path jobDir = temporaryFolder.newFolder("_job_dir").toPath();
+		final Path jobSubDir1 = Files.createDirectory(jobDir.resolve("_sub_dir1"));
+		final Path jobSubDir2 = Files.createDirectory(jobDir.resolve("_sub_dir2"));
+		final Path jarFile1 = Files.createFile(jobDir.resolve("file1.jar"));
+		final Path jarFile2 = Files.createFile(jobDir.resolve("file2.jar"));
+		final Path jarFile3 = Files.createFile(jobSubDir1.resolve("file3.jar"));
+		final Path jarFile4 = Files.createFile(jobSubDir2.resolve("file4.jar"));
+
+		Files.createFile(jobDir.resolve("file1.txt"));
+		Files.createFile(jobSubDir2.resolve("file2.txt"));
+
+		Path userDir = Paths.get(System.getProperty("user.dir"));
```

Comment:
Relying on system properties in a test is an indicator that the code should be refactored to make it more testable.